Indicators/AverageDirectionalIndex.cs

```diff
@@ -0,0 +1,332 @@
+                {
+                    postiveDirectionalMovement = input.High - _previousInput.High;
+                }
+            }
+
+            return postiveDirectionalMovement;
+        }
+
+        /// <summary>
+        /// Computes the negative directional movement.
+        /// </summary>
+        /// <param name="input">The input.</param>
+        /// <returns></returns>
+        private decimal ComputeNegativeDirectionalMovement(TradeBar input)
+        {
+            var negativeDirectionalMovement = new decimal(0.0);
+
+            if (_previousInput == null) return negativeDirectionalMovement;
+
+            if ((_previousInput.Low - input.Low) > (input.High - _previousInput.High))
+            {
+                if ((_previousInput.Low - input.Low) > 0)
+                {
+                    negativeDirectionalMovement = _previousInput.Low - input.Low;
+                }
+            }
+
+            return negativeDirectionalMovement;
+        }
+
+        /// <summary>
+        /// Computes the next value of this indicator from the given state
+        /// </summary>
+        /// <param name="input">The input given to the indicator</param>
+        /// <returns>A new value for this indicator</returns>
+        protected override decimal ComputeNextValue(TradeBar input)
+        {
+            TrueRange.Update(input);
+            DirectionalMovementPlus.Update(input);
+            DirectionalMovementMinus.Update(input);
+            SmoothedTrueRange.Update(Current);
+            SmoothedDirectionalMovementMinus.Update(Current);
+            SmoothedDirectionalMovementPlus.Update(Current);
+            PositiveDirectionalIndex.Update(Current);
+            NegativeDirectionalIndex.Update(Current);
+            var diff = Math.Abs(PositiveDirectionalIndex - NegativeDirectionalIndex);
+            var sum = PositiveDirectionalIndex + NegativeDirectionalIndex;
+            var value = Math.Round((sum == 0 ? 50 : ((_period - 1) * Current.Value + 100 * diff / sum) / _period), 5);
+            _previousInput = input;
+            return value;
+        }
+
+        /// <summary>
+        /// Computes the Plus Directional Indicator (+DI period).
+        /// </summary>
+        /// <returns></returns>
+        private decimal ComputePositiveDirectionalIndex()
+        {
+            if (SmoothedTrueRange == 0) return new decimal(0.0);
+
+            var positiveDirectionalIndex = (SmoothedDirectionalMovementPlus.Current.Value / TrueRange.Current.Value) * 100;
+
+            return positiveDirectionalIndex;
+        }
+
+        /// <summary>
+        /// Computes the Minus Directional Indicator (-DI period).
+        /// </summary>
+        /// <returns></returns>
+        private decimal ComputeNegativeDirectionalIndex()
+        {
+            if (SmoothedTrueRange == 0) return new decimal(0.0);
+
+            var negativeDirectionalIndex = (SmoothedDirectionalMovementMinus.Current.Value / TrueRange.Current.Value) * 100;
```

Comment:
Do we ever need to worry about TrueRange == 0m? I notice we check the SmoothedTrueRange, maybe it should b checking just the TrueRange?